<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>异常 | WangHao&amp;博客</title><meta name="author" content="WangHao"><meta name="copyright" content="WangHao"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="异常概念实际上，当java程序执行时出现错误时，jvm会把执行时出错的信息(例如出错原因、类型、 位置)收集，然后打包成为1个对象(object)，程序员可以对这种对象进行处理。这种对象就是所谓的异常。可能出现的异常的代码并不是肯定会出现异常, 还取决于执行环境和数据。 在Java中，所有的异常都有一个共同的超类Throwable（可抛出类）。Throwable类有两个重要的子类：Exceptio">
<meta property="og:type" content="article">
<meta property="og:title" content="异常">
<meta property="og:url" content="http://example.com/2021/04/29/%E5%BC%82%E5%B8%B8/index.html">
<meta property="og:site_name" content="WangHao&amp;博客">
<meta property="og:description" content="异常概念实际上，当java程序执行时出现错误时，jvm会把执行时出错的信息(例如出错原因、类型、 位置)收集，然后打包成为1个对象(object)，程序员可以对这种对象进行处理。这种对象就是所谓的异常。可能出现的异常的代码并不是肯定会出现异常, 还取决于执行环境和数据。 在Java中，所有的异常都有一个共同的超类Throwable（可抛出类）。Throwable类有两个重要的子类：Exceptio">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://example.com/img/40.jpg">
<meta property="article:published_time" content="2021-04-29T08:42:40.000Z">
<meta property="article:modified_time" content="2023-01-03T02:52:34.249Z">
<meta property="article:author" content="WangHao">
<meta property="article:tag" content="异常">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://example.com/img/40.jpg"><link rel="shortcut icon" href="/img/head.jpg"><link rel="canonical" href="http://example.com/2021/04/29/%E5%BC%82%E5%B8%B8/"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '异常',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-01-03 10:52:34'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.2.0"></head><body><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/img/head.jpg" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">101</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">120</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">13</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/img/40.jpg')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">WangHao&amp;博客</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> 分类</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> 归档</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> 标签</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> 关于</span></a></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">异常</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2021-04-29T08:42:40.000Z" title="发表于 2021-04-29 16:42:40">2021-04-29</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2023-01-03T02:52:34.249Z" title="更新于 2023-01-03 10:52:34">2023-01-03</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Java%E5%9F%BA%E7%A1%80/">Java基础</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="异常"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><h1 id="异常概念"><a href="#异常概念" class="headerlink" title="异常概念"></a>异常概念</h1><p>实际上，当java程序执行时出现错误时，jvm会把执行时出错的信息(例如出错原因、类型、 位置)收集，然后打包成为1个对象(object)，程序员可以对这种对象进行处理。这种对象就是所谓的异常。<strong>可能出现的异常的代码并不是肯定会出现异常, 还取决于执行环境和数据</strong>。</p>
<p>在Java中，所有的异常都有一个共同的超类Throwable（可抛出类）。Throwable类有两个重要的子类：Exception（异常）和 Error（错误），二者都是Java异常处理的重要子类，各自都包含大量子类。异常和错误的区别是：异常能被程序本身可以处理，错误是无法处理。</p>
<h1 id="异常分类：两种分类方法"><a href="#异常分类：两种分类方法" class="headerlink" title="异常分类：两种分类方法"></a>异常分类：两种分类方法</h1><h2 id="Error-和-Exception"><a href="#Error-和-Exception" class="headerlink" title="Error 和 Exception"></a>Error 和 Exception</h2><ul>
<li><strong>Error</strong>：表示程序在运行期间出现了非常严重的错误，并且该错误是不可恢复的，由于这属于JVM层次的严重错误，因此这种错误是会导致程序终止执行的。此外，编译器不会检查Error是否被处理，因此在程序中不推荐去捕获Error类型的异常，主要原因是运行时异常多是由于逻辑错误导致的，属于应该解决的错误，也就是说，一个正确的程序中是不应该存在Error的。OutOfMemoryError、ThreadDeath等都属于错误。当这些异常发生时，JVM一般会选择将线程终止。</li>
<li><strong>Exception</strong>：表示可恢复的异常，是编译器可以捕捉到的。可分为运行时异常（RuntimeException）和非运行时异常。</li>
</ul>
<span id="more"></span>

<h2 id="受查异常和不受查异常"><a href="#受查异常和不受查异常" class="headerlink" title="受查异常和不受查异常"></a>受查异常和不受查异常</h2><p>它包含两种类型：<strong>受查异常（checked exception）</strong>和<strong>不受查异常（unchecked exception）</strong>，其中不受查异常特指<strong>运行时异常（runtime exception）</strong>。</p>
<p>在实际开发中，经常会在程序编译时产生一些异常，而这些异常<strong>必须要进行处理</strong>，这种异常被称为<strong>编译时异常</strong>，也被称为<strong>受查异常</strong>，所有的Error都属于受查异常。另外还有一种异常是在程序运行时产生的（运行时异常），这种异常即使不编写异常处理代码，依然可以通过编译，因此被称为不受查异常。</p>
<p>总而言之，只能在代码中忽略RuntimeException（及其子类）类型的异常，其他类型异常的处理都是由编译器强制实施的。究其原因，RuntimeException代表的是编程错误：</p>
<ul>
<li>无法预料的错误。比如从你控制范围之外传递进来的null引用。</li>
<li>作为程序员，应该在代码中进行检查的错误。</li>
</ul>
<p>简而言之，它们之间的区别 ：</p>
<ul>
<li>非运行时异常：是需要显示用try-catch捕捉处理的异常，如IOException、SQLException等以及用户自定义的Exception异常。</li>
<li>运行时异常：不需要显示捕捉，如NullPointException，IndexOutOfBoundsException等，常见于程序逻辑问题导致。</li>
</ul>
<h1 id="异常处理机制"><a href="#异常处理机制" class="headerlink" title="异常处理机制"></a>异常处理机制</h1><h2 id="捕获与处理"><a href="#捕获与处理" class="headerlink" title="捕获与处理"></a>捕获与处理</h2><h3 id="try-catch-finally的处理机制"><a href="#try-catch-finally的处理机制" class="headerlink" title="try-catch-finally的处理机制"></a>try-catch-finally的处理机制</h3><p>完整的异常处理程序看起来像这样，但Java也允许缺少catch或者finally：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    <span class="comment">// 监控区域（guarded region）</span></span><br><span class="line">    <span class="comment">// 可能抛出异常 A、B、C 的危险活动</span></span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">A</span>(); <span class="comment">// A为自定义的异常类型</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(A a1) &#123;</span><br><span class="line">    <span class="comment">// 异常A的异常处理程序</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(B b1) &#123;</span><br><span class="line">    <span class="comment">// 异常B的异常处理程序</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(C c1) &#123;</span><br><span class="line">    <span class="comment">// 异常C的异常处理程序</span></span><br><span class="line">&#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">    <span class="comment">// 每次必须执行的活动</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<ul>
<li><strong>try</strong>块为监控区域（guarded region），即可能产生异常的区域。关键字throw用来抛出异常，每个异常都作为一个对象抛出。</li>
<li><strong>catch</strong>块为异常处理程序，异常处理程序必须紧跟在try块之后，当异常被抛出时，异常处理程序将负责搜寻参数与异常类型相匹配的第一个处理程序，然后进入catch子句执行。注意，只有匹配的catch子句才能得到执行（第5部分将会讲到异常匹配），这与switch语句不同，switch语句需要在每一个case后面跟一个break，以避免执行后续的case子句。</li>
<li><strong>finally</strong>块是程序希望每次都要执行的代码，例如一些需要清理的资源：已经打开的文件或网络连接，在屏幕上画的图形，甚至可以是外部世界的某个开关。总之记住，finally最可靠。</li>
</ul>
<h3 id="自动关闭的try-catch"><a href="#自动关闭的try-catch" class="headerlink" title="自动关闭的try-catch"></a>自动关闭的try-catch</h3><p>补充：在Java7之前，如果遇到类似IO流的处理，要在finally部分关闭IO流；Java7增加了一个新用法，在Java7优化后的try-with-resource语句，该语句确保了每个资源在语句结束时关闭。所谓的资源是指在程序完成后，必须关闭的流对象。写在()里面的流对象对应的类都实现了自动关闭接口AutoCloseable；</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">try (创建流对象语句，如果多个,使用&#x27;;&#x27;隔开) &#123;</span><br><span class="line">    // 读写数据</span><br><span class="line">&#125; catch (IOException e) &#123;</span><br><span class="line">    e.printStackTrace();</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h3 id="try-catch-finally的执行机制"><a href="#try-catch-finally的执行机制" class="headerlink" title="try-catch-finally的执行机制"></a>try-catch-finally的执行机制</h3><p>一般的顺序就是try-&gt;catch-&gt;finally，但要注意当出现return时的情况。若在try或catch语句里面有return语句，finally语句和return语句的执行顺序问题：</p>
<ul>
<li><p>若有finally语句，则无论如何，都会执行该语句，在try或catch中的return语句会将它的返回值压入栈内，然后执行finally语句，当finally执行完成后，若finally语句里有return语句，则执行return语句并结束。return的作用就是结束当前方法。</p>
</li>
<li><p>若finally没有return语句，则返回被保存的栈里的return语句，再执行。然而，在压栈时候，要注意压入栈内的是什么东西，是值本身还是引用，若是引用则引用的值会改变，若是变量值，则不会改变。然而非常不建议在finally里面写return语句，因为它会吞掉异常，后面有例子。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">Multiplereturns</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">f</span><span class="params">(<span class="type">int</span> i)</span> &#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;Initialization that requires cleadup&quot;</span>);</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Point 1&quot;</span>);</span><br><span class="line">            <span class="keyword">if</span> (i == <span class="number">1</span>)</span><br><span class="line">                <span class="keyword">return</span>;</span><br><span class="line">            System.out.println(<span class="string">&quot;Point 2&quot;</span>);</span><br><span class="line">            <span class="keyword">if</span> (i == <span class="number">2</span>)</span><br><span class="line">                <span class="keyword">return</span>;</span><br><span class="line">            System.out.println(<span class="string">&quot;Point 3&quot;</span>);</span><br><span class="line">            <span class="keyword">if</span> (i == <span class="number">3</span>)</span><br><span class="line">                <span class="keyword">return</span>;</span><br><span class="line">            System.out.println(<span class="string">&quot;End&quot;</span>);</span><br><span class="line">            <span class="keyword">return</span>;</span><br><span class="line">        &#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Performing cleanup&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        <span class="keyword">for</span> (<span class="type">int</span> <span class="variable">i</span> <span class="operator">=</span> <span class="number">1</span>; i &lt;= <span class="number">4</span>; i++)</span><br><span class="line">            f(i);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">Initialization that requires cleadup</span></span><br><span class="line"><span class="comment">Point 1</span></span><br><span class="line"><span class="comment">Performing cleanup</span></span><br><span class="line"><span class="comment">Initialization that requires cleadup</span></span><br><span class="line"><span class="comment">Point 1</span></span><br><span class="line"><span class="comment">Point 2</span></span><br><span class="line"><span class="comment">Performing cleanup</span></span><br><span class="line"><span class="comment">Initialization that requires cleadup</span></span><br><span class="line"><span class="comment">Point 1</span></span><br><span class="line"><span class="comment">Point 2</span></span><br><span class="line"><span class="comment">Point 3</span></span><br><span class="line"><span class="comment">Performing cleanup</span></span><br><span class="line"><span class="comment">Initialization that requires cleadup</span></span><br><span class="line"><span class="comment">Point 1</span></span><br><span class="line"><span class="comment">Point 2</span></span><br><span class="line"><span class="comment">Point 3</span></span><br><span class="line"><span class="comment">End</span></span><br><span class="line"><span class="comment">Performing cleanup</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure></li>
</ul>
<h3 id="异常丢失"><a href="#异常丢失" class="headerlink" title="异常丢失"></a>异常丢失</h3><p>当用特殊的方式使用finally，可能会造成异常丢失，Thinking in Java中举了两个例子，如下：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">VeryImportantException</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> String <span class="title function_">toString</span><span class="params">()</span> &#123;</span><br><span class="line">        <span class="keyword">return</span> <span class="string">&quot;A very important exception!&quot;</span>;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">HoHumException</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> String <span class="title function_">toString</span><span class="params">()</span> &#123;</span><br><span class="line">        <span class="keyword">return</span> <span class="string">&quot;A trivial exception&quot;</span>;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">LostMessage</span> &#123;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> <span class="keyword">throws</span> VeryImportantException &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">VeryImportantException</span>();</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">dispose</span><span class="params">()</span> <span class="keyword">throws</span> HoHumException &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">HoHumException</span>();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="type">LostMessage</span> <span class="variable">lm</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">LostMessage</span>();</span><br><span class="line">            <span class="keyword">try</span> &#123;</span><br><span class="line">                lm.f();</span><br><span class="line">            &#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">                lm.dispose();</span><br><span class="line">            &#125;</span><br><span class="line">        &#125; <span class="keyword">catch</span>(Exception e) &#123;</span><br><span class="line">            System.out.println(e);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/* Output:</span></span><br><span class="line"><span class="comment">A trivial exception</span></span><br><span class="line"><span class="comment">*/</span><span class="comment">//:~</span></span><br></pre></td></tr></table></figure>

<p>从输出可以看到，VeryImportantException不见了，在这种嵌套的异常处理机制中，程序对lm.f()所抛出的异常没有做任何处理，但却骗过了编译器。内层产生了两个异常，因为catch只能捕获一个异常，而编译器默认内层只能产生一个异常，所以它觉得自己已经对异常进行了处理。而实际上对lm.f()产生的异常没有任何处理但却通过了编译，这是一种非常可怕的缺陷。</p>
<p>注意此种情况与下文将要讲的<strong>重抛异常中的异常链</strong>的区别，重抛异常是在catch里面，即已经对上一个异常进行了处理，只是为了保留上一个异常的信息，不存在编译错误。<br>还有一种异常丢失的情况是在没有catch或者产生catch不匹配的异常的情况下就在fianlly子句中返回：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">public class ExceptionSilencer &#123;</span><br><span class="line">    public static void main(String[] args) &#123;</span><br><span class="line">        try &#123;</span><br><span class="line">            throw new RuntimeException();</span><br><span class="line">        &#125; finally &#123;</span><br><span class="line">            // 在finally里返回，会忽略掉所有抛出的异常</span><br><span class="line">            return;</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h2 id="捕获所有异常"><a href="#捕获所有异常" class="headerlink" title="捕获所有异常"></a>捕获所有异常</h2><h3 id="异常的信息：e-toString-和e-getMessage"><a href="#异常的信息：e-toString-和e-getMessage" class="headerlink" title="异常的信息：e.toString()和e.getMessage()"></a>异常的信息：e.toString()和e.getMessage()</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">public class E03 &#123;</span><br><span class="line">    public static void main(String[] args) &#123;</span><br><span class="line">        try&#123;</span><br><span class="line">            throw new ArrayIndexOutOfBoundsException(&quot;An ArrayIndexOutOfBoundsException in main&quot;);</span><br><span class="line">            //throw new ArrayIndexOutOfBoundsException();</span><br><span class="line">        &#125;</span><br><span class="line">        catch (ArrayIndexOutOfBoundsException e) &#123;</span><br><span class="line">            System.out.println(e.getMessage());</span><br><span class="line">            System.out.println(e.toString());</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line">/*Output:</span><br><span class="line">An ArrayIndexOutOfBoundsException in main</span><br><span class="line">java.lang.ArrayIndexOutOfBoundsException: An ArrayIndexOutOfBoundsException in main</span><br><span class="line">-----------------------------------------------------------------------------------</span><br><span class="line">null</span><br><span class="line">java.lang.ArrayIndexOutOfBoundsException</span><br><span class="line"> */</span><br></pre></td></tr></table></figure>

<p>上面是传入字符串的输出，下面是默认构造器的输出。可以初步判断e.getMessage()获取的是传入的信息，而e.toString()包含两部分：类的信息+传入的信息。是否是这样哪？看源码：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">private String detailMessage;</span><br><span class="line"></span><br><span class="line">public Throwable(String message) &#123;</span><br><span class="line">    fillInStackTrace();</span><br><span class="line">    detailMessage = message;</span><br><span class="line">&#125;</span><br><span class="line">...Throwable的构造方法都接收message传入detailMessage</span><br><span class="line"></span><br><span class="line">public String getMessage() &#123;</span><br><span class="line">    return detailMessage;</span><br><span class="line">&#125;</span><br><span class="line">...getMessage()获取detailMessage</span><br><span class="line"></span><br><span class="line">public String toString() &#123;</span><br><span class="line">    String s = getClass().getName();</span><br><span class="line">    String message = getLocalizedMessage();</span><br><span class="line">    return (message != null) ? (s + &quot;: &quot; + message) : s;</span><br><span class="line">&#125;</span><br><span class="line">public String getLocalizedMessage() &#123;</span><br><span class="line">    return getMessage();</span><br><span class="line">&#125;</span><br><span class="line">...toString()包含两部分：类的信息+传入的信息Message，真相大白了！</span><br></pre></td></tr></table></figure>

<p>所以通过源码我们可以看出信息的详细程度为：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">getMessage() &lt;= getLocalizedMessage() &lt; toString() &lt; printStackTrace()</span><br></pre></td></tr></table></figure>

<h3 id="栈轨迹和重抛异常"><a href="#栈轨迹和重抛异常" class="headerlink" title="栈轨迹和重抛异常"></a>栈轨迹和重抛异常</h3><h4 id="e-printStackTrace"><a href="#e-printStackTrace" class="headerlink" title="e.printStackTrace()"></a>e.printStackTrace()</h4><p>首先介绍大家常用的e.printStackTrace()，printStackTrace方法源码包含3个重载的方法，其中第一个方法输出到标准错误，后两个允许选择要输出的流。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title function_">printStackTrace</span><span class="params">()</span> &#123;printStackTrace(System.err);&#125;</span><br><span class="line"><span class="comment">/**</span></span><br><span class="line"><span class="comment"> * Prints this throwable and its backtrace to the specified print stream.</span></span><br><span class="line"><span class="comment"> *</span></span><br><span class="line"><span class="comment"> * <span class="doctag">@param</span> s &#123;<span class="doctag">@code</span> PrintStream&#125; to use for output</span></span><br><span class="line"><span class="comment"> */</span></span><br><span class="line"> </span><br><span class="line"><span class="keyword">public</span> <span class="keyword">void</span> <span class="title function_">printStackTrace</span><span class="params">(PrintStream s)</span> &#123;</span><br><span class="line">    printStackTrace(<span class="keyword">new</span> <span class="title class_">WrappedPrintStream</span>(s));</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">private</span> <span class="keyword">void</span> <span class="title function_">printStackTrace</span><span class="params">(PrintStreamOrWriter s)</span> &#123;...&#125;</span><br></pre></td></tr></table></figure>

<p>printStackTrace()方法将此Throwable对象的堆栈跟踪输出至错误输出流，作为字段 System.err 的值。输出的第一行包含此对象的toString()方法的结果；剩余行表示由方法 <strong>fillInStackTrace()</strong> （下面将详细介绍）记录的数据，即Throwable的调用栈轨迹。</p>
<p>调用栈轨迹简单理解就是将调用的方法序列存储在栈中，printStackTrace()方法提供的信息可以通过getStackTrace()方法来直接访问，这个方法将返回栈轨迹中的元素所构成的数组。<strong>其中每一个元素都表示栈中的一帧，栈顶是元素0，也是调用序列最后一个方法调用（异常被创建和抛出之处）；栈底是调用序列中的第一个方法调用</strong>。例如：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">WhoCalled</span> &#123;</span><br><span class="line">    <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> &#123;</span><br><span class="line">        <span class="comment">// Generate an exception to fill in the stack trace</span></span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Exception</span>();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            <span class="keyword">for</span> (StackTraceElement ste : e.getStackTrace())</span><br><span class="line">                System.out.println(ste.getMethodName());</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">g</span><span class="params">()</span> &#123; f(); &#125;</span><br><span class="line">    <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">h</span><span class="params">()</span> &#123; g(); &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        f();</span><br><span class="line">        System.out.println(<span class="string">&quot;-----------------&quot;</span>);</span><br><span class="line">        g();</span><br><span class="line">        System.out.println(<span class="string">&quot;-----------------&quot;</span>);</span><br><span class="line">        h();</span><br><span class="line">        System.out.println(<span class="string">&quot;-----------------&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">f</span></span><br><span class="line"><span class="comment">main</span></span><br><span class="line"><span class="comment">-----------------</span></span><br><span class="line"><span class="comment">f</span></span><br><span class="line"><span class="comment">g</span></span><br><span class="line"><span class="comment">main</span></span><br><span class="line"><span class="comment">-----------------</span></span><br><span class="line"><span class="comment">f</span></span><br><span class="line"><span class="comment">g</span></span><br><span class="line"><span class="comment">h</span></span><br><span class="line"><span class="comment">main</span></span><br><span class="line"><span class="comment">-----------------</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<p>第一个方法调用是main()方法，首先入栈，在栈底；创建异常的地方在f()，最后入栈，也就是栈顶。</p>
<h4 id="重新抛出异常"><a href="#重新抛出异常" class="headerlink" title="重新抛出异常"></a>重新抛出异常</h4><p>有时希望把刚捕获的异常重新抛出，尤其是在使用Exception捕获所有异常时，既然已经得到了对当前异常对象的引用，可以直接把它重新抛出。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">catch(Exception e) &#123;</span><br><span class="line">    System.out.printin(&quot;An exception was thrown&quot;);</span><br><span class="line">    throw e;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><strong>重抛异常会把异常抛给上一级环境中的异常处理程序，同一个try块的后续catch子句将被忽略。如果只是把当前异常对象重新抛出，那么 printStackTrace方法显示的将是原来异常抛出点的调用栈信息，而并非重新抛出点的信息</strong>。要想更新这个信息，可以调用fillInStackTrace()方法，将返回一个Throwable对象，它是通过把当前调用栈信息填入原来那个异常対象而建立的，像这样：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">Rethrowing</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> <span class="keyword">throws</span> Exception &#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;originating the exception in f()&quot;</span>);</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Exception</span>(<span class="string">&quot;thrown from f()&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">g</span><span class="params">()</span> <span class="keyword">throws</span> Exception &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            f();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Inside g(),e.printStackTrace&quot;</span>);</span><br><span class="line">            e.printStackTrace();</span><br><span class="line">            System.out.println(<span class="string">&quot;-----------------------------------------------------&quot;</span>);</span><br><span class="line">            <span class="keyword">throw</span> e; <span class="comment">//rethrow</span></span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">h</span><span class="params">()</span> <span class="keyword">throws</span> Exception &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            f();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Inside h(),e.printStackTrace&quot;</span>);</span><br><span class="line">            e.printStackTrace();</span><br><span class="line">            System.out.println(<span class="string">&quot;-----------------------------------------------------&quot;</span>);</span><br><span class="line">            <span class="keyword">throw</span> (Exception) e.fillInStackTrace(); <span class="comment">//rethrow</span></span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            g();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;main: printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">            System.out.println(<span class="string">&quot;-----------------------------------------------------&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            h();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;main: printStackTrace()&quot;</span>);</span><br><span class="line">            e.printStackTrace(System.out);</span><br><span class="line">            System.out.println(<span class="string">&quot;-----------------------------------------------------&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*</span></span><br><span class="line"><span class="comment">originating the exception in f()</span></span><br><span class="line"><span class="comment">Inside g(),e.printStackTrace</span></span><br><span class="line"><span class="comment">java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.f(Rethrowing.java:6)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.g(Rethrowing.java:10)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.main(Rethrowing.java:33)</span></span><br><span class="line"><span class="comment">-----------------------------------------------------</span></span><br><span class="line"><span class="comment">main: printStackTrace()</span></span><br><span class="line"><span class="comment">java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.f(Rethrowing.java:6)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.g(Rethrowing.java:10)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.main(Rethrowing.java:33)</span></span><br><span class="line"><span class="comment">-----------------------------------------------------</span></span><br><span class="line"><span class="comment">originating the exception in f()</span></span><br><span class="line"><span class="comment">Inside h(),e.printStackTrace</span></span><br><span class="line"><span class="comment">java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.f(Rethrowing.java:6)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.h(Rethrowing.java:21)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.main(Rethrowing.java:41)</span></span><br><span class="line"><span class="comment">-----------------------------------------------------</span></span><br><span class="line"><span class="comment">main: printStackTrace()</span></span><br><span class="line"><span class="comment">java.lang.Exception: thrown from f()</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.h(Rethrowing.java:27)</span></span><br><span class="line"><span class="comment">   at chapter12.Rethrowing.main(Rethrowing.java:41)</span></span><br><span class="line"><span class="comment">-----------------------------------------------------</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<p>调用fillInStackTrace()那一行就成为异常的新发生地了，如果是在捕获异常之后抛出新异常，得到的效果类似于使用fillInStackTrace()，即得到新的抛出点相关的信息。java中e.printStackTrace()请谨慎使用，最好使用logger记录，可能会导致锁死。或者建议使用：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">logger.error(&quot;异常信息&quot;,e);//注意这个方法在java.util.logging.Logger类里没有</span><br></pre></td></tr></table></figure>

<h3 id="异常链"><a href="#异常链" class="headerlink" title="异常链"></a>异常链</h3><ul>
<li><p>常常会再捕获一个异常后抛出另外一个异常，并且希望把<strong>异常原始信息保存下来，这被称为异常链</strong>。</p>
</li>
<li><p>现在所有Throwable的子类在构造器中都可以接受一个cause（因由）对象作为参数，这个cause表示原始异常，这样通过把原始异常传递给新的异常，使得即使在当前位置创建并抛出了新的异常，也可以通过这个cause追踪到异常最初发生的位置。</p>
</li>
<li><p>Throwable类及其所有的子类都提供了带cause参数的构造器，其他的异常类就只有通过initCause()来设置cause了。设置完cause的新异常可以通过getCause()方法返回原始异常。</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">ExceptionCause</span> &#123;</span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">test1</span><span class="params">()</span> <span class="keyword">throws</span> Exception&#123;</span><br><span class="line">        <span class="keyword">try</span>&#123;</span><br><span class="line">            test2();                                      <span class="comment">// 创建原始异常</span></span><br><span class="line">        &#125;<span class="keyword">catch</span>(NullPointerException e)&#123;                   <span class="comment">// 捕获原始异常</span></span><br><span class="line">            <span class="type">Exception</span> <span class="variable">Ex</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">Exception</span>(<span class="string">&quot;New exception&quot;</span>);<span class="comment">// 1</span></span><br><span class="line">            Ex.initCause(e);                              <span class="comment">// 把原始异常传递给新的异常</span></span><br><span class="line">            <span class="keyword">throw</span> Ex;                                     <span class="comment">// 抛出新的异常</span></span><br><span class="line"> <span class="comment">//           Ex.getCause();                                // 原始异常</span></span><br><span class="line"><span class="comment">//           throw new Exception(&quot;New exception&quot;, e);      // 2</span></span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">test2</span><span class="params">()</span>&#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span>  <span class="title class_">NullPointerException</span>(<span class="string">&quot;Str is null&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> <span class="keyword">throws</span> Exception &#123;</span><br><span class="line">        test1();</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">Exception in thread &quot;main&quot; java.lang.Exception: New exception</span></span><br><span class="line"><span class="comment">   at chapter12.ExceptionCause.test1(ExceptionCause.java:8)</span></span><br><span class="line"><span class="comment">   at chapter12.ExceptionCause.main(ExceptionCause.java:19)</span></span><br><span class="line"><span class="comment">Caused by: java.lang.NullPointerException: Str is null       //如果没有initCause(e),原始异常不会输出</span></span><br><span class="line"><span class="comment">   at chapter12.ExceptionCause.test2(ExceptionCause.java:15)</span></span><br><span class="line"><span class="comment">   at chapter12.ExceptionCause.test1(ExceptionCause.java:6)</span></span><br><span class="line"><span class="comment">   ... 1 more</span></span><br><span class="line"><span class="comment">*/</span>  </span><br></pre></td></tr></table></figure></li>
</ul>
<h1 id="自定义异常"><a href="#自定义异常" class="headerlink" title="自定义异常"></a>自定义异常</h1><h2 id="创建自定义异常"><a href="#创建自定义异常" class="headerlink" title="创建自定义异常"></a>创建自定义异常</h2><p>要自己定义异常类，必须从已有的异常类继承，最好是选择意思相近的异常类（但是并不好找）。建立新的异常类非常简单，可以是简单的继承，例如：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">class MyException extends Exception &#123;&#125;</span><br></pre></td></tr></table></figure>

<p>或者你想加点更详细的信息：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">MyException</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;</span><br><span class="line">    String s;</span><br><span class="line">    MyException(String s) &#123;</span><br><span class="line">        <span class="built_in">super</span>(s);  <span class="comment">//可以用getMessage了</span></span><br><span class="line">        <span class="built_in">this</span>.s = s;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">public</span> String <span class="title function_">toString</span><span class="params">()</span> &#123;</span><br><span class="line">        <span class="keyword">return</span> s;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">E04</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">MyException</span>(<span class="string">&quot;A MyException in main&quot;</span>);</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (MyException m) &#123;</span><br><span class="line">            <span class="comment">//温习一下前面的知识</span></span><br><span class="line">            System.out.println(m.getMessage());</span><br><span class="line">            System.out.println(m.toString());</span><br><span class="line">            m.printStackTrace();</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">A MyException in main</span></span><br><span class="line"><span class="comment">A MyException in main</span></span><br><span class="line"><span class="comment">A MyException in main</span></span><br><span class="line"><span class="comment">A MyException in main</span></span><br><span class="line"><span class="comment">	at chapter12.exercise4.E04.main(E04.java:17)</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<h2 id="异常说明"><a href="#异常说明" class="headerlink" title="异常说明"></a>异常说明</h2><p>Java鼓励人们把方法可能会抛出的异常告知使用此方法的客户端程序员，这是种优雅的做法，它使得调用者能确切知道写什么样的代码可以捕获所有潜在的异常。当然，如果提供了源代码，客户端程序员可以在源代码中查找throw语句来获知相关信息，然而程序库通常不与源代码一起发布。<strong>为了预防这样的问题，java提供了相应的语法来告知程序员方法可能会抛出的异常类型，这就是异常说明，属于方法声明的一部分，紧跟在形式参数列表之后</strong>。</p>
<p>异常说明可能有两种意思，一个是“我的代码会产生这种异常，这由你来处理”。另一个是“我的代码忽略了这些异常，这由你来处理”。学习异常处理的机制和语法的时候，我们一直在关注“你来处理”部分，但这里特别值得注意的事实是我们通常都忽略了异常说明所表达的完整含义。</p>
<p>异常说明使用了附加的关键字 throws，后面接一个所有潜在异常类型的列表，所以方法可能看起来像这样：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">void f() throws Exception1,Exception2,Exception3 &#123;&#125;</span><br></pre></td></tr></table></figure>

<p>但是，要是这样写：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">void f() &#123;&#125;</span><br></pre></td></tr></table></figure>

<p>就表示此方法不会抛出任何异常(<strong>除了从 RuntimeException继承的异常，它们可以在没有异常说明的情况下被抛出</strong>)。<br>例如：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">E08</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> <span class="keyword">throws</span> MyException&#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">MyException</span>();</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">g</span><span class="params">()</span> &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">NullPointerException</span>();</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            E08.f();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            e.printStackTrace();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            E08.g();</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">catch</span> (Exception e) &#123;</span><br><span class="line">            e.printStackTrace();</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">chapter12.exercise8.MyException</span></span><br><span class="line"><span class="comment">   at chapter12.exercise8.E08.f(E08.java:7)</span></span><br><span class="line"><span class="comment">   at chapter12.exercise8.E08.main(E08.java:14)</span></span><br><span class="line"><span class="comment">java.lang.NullPointerException</span></span><br><span class="line"><span class="comment">   at chapter12.exercise8.E08.g(E08.java:10)</span></span><br><span class="line"><span class="comment">   at chapter12.exercise8.E08.main(E08.java:20)</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<p>上例必须在f()方法里进行异常声明，否则编译器会报错，因为自定义的异常属于非运行时异常，而NullPointerException属于运行时异常。</p>
<p><strong>代码必须与异常说明保持一致</strong>。如果方法里的代码产生了异常（非运行时异常）却没有处理，编译器会发现这个问题并提醒你：要么处理这个异常，要么在异常说明中说明此方法将产生异常。</p>
<p><strong>不过还是有个能“作弊”的地方：可以声明方法将抛出异常，实际上却不抛出，编译器相信了这个声明，并强制使用此方法的用户像真的抛出异常那样使用这个方法</strong>。这样做的好处是为异常先占个位子，以后就可以抛出这种异常而不用修改已有的代码。<strong>在定义抽象基类和接口时这种能力很重要，这样派生类或接口实现就能够抛出这些预先声明的异常。</strong></p>
<h2 id="异常的限制"><a href="#异常的限制" class="headerlink" title="异常的限制"></a>异常的限制</h2><p>但要注意当覆盖方法的时候，只能抛出在基类方法的异常说明中列出的那些异常（包括其子类）。这个限制意味着当基类使用的代码应用到其派生类对象的时候，一样能够工作（这是面向对象的概念）异常也不例外。例如：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">E1</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">E2</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">E1_1</span> <span class="keyword">extends</span> <span class="title class_">E1</span> &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">interface</span> <span class="title class_">A</span> &#123;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">test</span><span class="params">()</span> <span class="keyword">throws</span> E1;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">abstract</span> <span class="keyword">class</span> <span class="title class_">B</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">void</span> <span class="title function_">test</span><span class="params">()</span> <span class="keyword">throws</span> E2&#123;&#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Scratch</span> <span class="keyword">extends</span> <span class="title class_">B</span> <span class="keyword">implements</span> <span class="title class_">A</span> &#123;&#125;</span><br></pre></td></tr></table></figure>

<p>此时无法通过编译，因为Scratch类继承了B的方法作为接口方法的实现，但E2却不是E1及其子类。如果将类B换成如下即可：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">abstract</span> <span class="keyword">class</span> <span class="title class_">B</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">void</span> <span class="title function_">test</span><span class="params">()</span> <span class="keyword">throws</span> E1&#123;&#125;</span><br><span class="line"><span class="comment">//    public void test() throws E1_1&#123;&#125;</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>或者在Scratch里实现test，但test不能抛出任何异常，此时类B中的test()方法与接口A中的test()方法没有直接关系：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">Scratch</span> <span class="keyword">extends</span> <span class="title class_">B</span> <span class="keyword">implements</span> <span class="title class_">A</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">void</span> <span class="title function_">test</span><span class="params">()</span> &#123;&#125; </span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h2 id="异常与记录日志"><a href="#异常与记录日志" class="headerlink" title="异常与记录日志"></a>异常与记录日志</h2><h3 id="System-err和System-out的区别（补充）"><a href="#System-err和System-out的区别（补充）" class="headerlink" title="System.err和System.out的区别（补充）"></a>System.err和System.out的区别（补充）</h3><p>官方文档中给出的解释如下：</p>
<ul>
<li><p>System.out：“标准”输出流。通常，此流对应于显示器输出或者由主机环境或用户指定的另一个输出目标。对于简单独立的 Java 应用程序，编写一行输出数据的典型方式是：System.out.println(data)；</p>
</li>
<li><p>System.err：“标准”错误输出流。通常，此流对应于显示器输出或者由主机环境或用户指定的另一个输出目标。按照惯例，此输出流用于显示错误消息，或者显示那些即使用户输出流（变量 out 的值）已经重定向到通常不被连续监视的某一文件或其他目标，也应该立刻引起用户注意的其他信息。</p>
</li>
<li><p>区别：字面上理解二者的区别即是out用于正常的用户指定的输出，而err专门用于显示错误和异常消息；System.err输出的字符串位置会随机出现，但System.err输出的字符串之间的相对位置不会改变，System.out在JVM和操作系统都具有缓存功能，就是你输出的东西不一定实时输出，有时候会积攒到一定数量才会输出，System.err会实时输出，不需要缓冲，单独使用的话可能感觉不到，如果两种方式混合使用就会发现了。</p>
<p>例如：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">PrintTset</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String args[])</span> &#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;out1&quot;</span>);</span><br><span class="line">        System.err.println(<span class="string">&quot;err1&quot;</span>);</span><br><span class="line">        System.out.println(<span class="string">&quot;out2&quot;</span>);</span><br><span class="line">        System.err.println(<span class="string">&quot;err2&quot;</span>);</span><br><span class="line">        System.out.println(<span class="string">&quot;out3&quot;</span>);</span><br><span class="line">        System.out.println(<span class="string">&quot;out4&quot;</span>);</span><br><span class="line">        System.out.println(<span class="string">&quot;out5&quot;</span>);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">out1</span></span><br><span class="line"><span class="comment">err1</span></span><br><span class="line"><span class="comment">err2</span></span><br><span class="line"><span class="comment">out2</span></span><br><span class="line"><span class="comment">out3</span></span><br><span class="line"><span class="comment">out4</span></span><br><span class="line"><span class="comment">out5</span></span><br><span class="line"><span class="comment">---</span></span><br><span class="line"><span class="comment">out1</span></span><br><span class="line"><span class="comment">err1</span></span><br><span class="line"><span class="comment">out2</span></span><br><span class="line"><span class="comment">err2</span></span><br><span class="line"><span class="comment">out3</span></span><br><span class="line"><span class="comment">out4</span></span><br><span class="line"><span class="comment">out5</span></span><br><span class="line"><span class="comment">---</span></span><br><span class="line"><span class="comment">...</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<p>输出结果每次都不一样，但System.err和System.out输出的相对顺序保持不变，这也是由于System.err不需要缓冲即可输出造成的。</p>
</li>
</ul>
<h3 id="Java-Logger使用"><a href="#Java-Logger使用" class="headerlink" title="Java Logger使用"></a>Java Logger使用</h3><p>…</p>
<h1 id="异常深入"><a href="#异常深入" class="headerlink" title="异常深入"></a>异常深入</h1><h2 id="构造器"><a href="#构造器" class="headerlink" title="构造器"></a>构造器</h2><p>有一点很重要，即你要时刻询问自己“如果异常发生了，所有东西能被正确的清理吗?”尽管大多数情况下是非常安全的，但涉及构造器时，问题就出现了，构造器会把对象设置成安全的初始状态，但还会有别的动作，比如打开一个文件，这样的动作只有在对象使用完毕并且用户调用了特殊的清理方法之后才能得以清理。如果在构造器内抛出了异常，这些清理行为也许就不能正常工作了。这意味着在编写构造器时要格外细心。</p>
<p>finally能解决所有问题吗？因为finally会每次执行清理代码，但如果构造器在其执行过程中半途而废，也许该对象的某些部分还没有被成功创建，而这些部分却是在finally里要清理的，这就出现问题了。</p>
<p>对于在构造阶段可能会抛出异常，并且要求清理的类，最安全的使用方式是使用嵌套的try子句，这种通用的清理惯用法在构造器不抛出异常时也应该运用，其基本规则是：<strong>在创建需要清理的对象之后，立即进入一个try-finally的语句块</strong>：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">ConstructionException</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">WithDispose</span> &#123;</span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">final</span> <span class="type">int</span> id;</span><br><span class="line">    WithDispose(<span class="type">int</span> id) &#123;</span><br><span class="line">        <span class="built_in">this</span>.id = id;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">dispose</span><span class="params">()</span> &#123;</span><br><span class="line">        System.out.println(<span class="string">&quot;WithDispose.dispose(): &quot;</span> + id);</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">FailingConstructor</span> &#123;</span><br><span class="line">    <span class="keyword">private</span> <span class="keyword">final</span> WithDispose wd1, wd2;</span><br><span class="line">    <span class="comment">// 构造器可能在创建过程中失败并且抛出一个异常</span></span><br><span class="line">    FailingConstructor(<span class="type">boolean</span> fail) <span class="keyword">throws</span> ConstructionException &#123;</span><br><span class="line">        wd1 = <span class="keyword">new</span> <span class="title class_">WithDispose</span>(<span class="number">1</span>);</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="comment">// &quot;Simulate&quot; other code that might throw exceptions</span></span><br><span class="line">             <span class="keyword">if</span>(fail) <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">ConstructionException</span>();</span><br><span class="line">        &#125; <span class="keyword">catch</span>(ConstructionException e) &#123;</span><br><span class="line">            wd1.dispose();</span><br><span class="line">            <span class="keyword">throw</span> e;</span><br><span class="line">        &#125;</span><br><span class="line">        wd2 = <span class="keyword">new</span> <span class="title class_">WithDispose</span>(<span class="number">2</span>);</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="comment">// 用于清理成员对象的方法</span></span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">dispose</span><span class="params">()</span> &#123;   </span><br><span class="line">        wd2.dispose();</span><br><span class="line">        wd1.dispose();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">E24_FailingConstructor</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String args[])</span> &#123;</span><br><span class="line">        <span class="keyword">for</span>(<span class="type">boolean</span> <span class="variable">b</span> <span class="operator">=</span> <span class="literal">false</span>; ; b = !b)</span><br><span class="line">            <span class="keyword">try</span> &#123;</span><br><span class="line">                System.out.println(<span class="string">&quot;Constructing...&quot;</span>);</span><br><span class="line">                <span class="type">FailingConstructor</span> <span class="variable">fc</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">FailingConstructor</span>(b);</span><br><span class="line">                <span class="comment">// 在创建需要清理的对象之后，立即进入一个try-finally的语句块</span></span><br><span class="line">                <span class="comment">// 构造过程中抛出异常会终止进入外层catch，否则正常运行内层的try-finally块</span></span><br><span class="line">                <span class="keyword">try</span> &#123;</span><br><span class="line">                    System.out.println(<span class="string">&quot;Processing...&quot;</span>);</span><br><span class="line">                &#125; <span class="keyword">finally</span> &#123;</span><br><span class="line">                    System.out.println(<span class="string">&quot;Cleanup...&quot;</span>);</span><br><span class="line">                    fc.dispose();  <span class="comment">// 记得清理，finally在内层就可以正常发挥作用了</span></span><br><span class="line">                    System.out.println(<span class="string">&quot;-----------------------------&quot;</span>);</span><br><span class="line">                &#125;</span><br><span class="line">            &#125; <span class="keyword">catch</span>(ConstructionException e) &#123;</span><br><span class="line">                System.out.println(<span class="string">&quot;Construction has failed: &quot;</span> + e);</span><br><span class="line">                <span class="keyword">break</span>;</span><br><span class="line">            &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">Constructing...</span></span><br><span class="line"><span class="comment">Processing...</span></span><br><span class="line"><span class="comment">Cleanup...</span></span><br><span class="line"><span class="comment">WithDispose.dispose(): 2</span></span><br><span class="line"><span class="comment">WithDispose.dispose(): 1</span></span><br><span class="line"><span class="comment">-----------------------------</span></span><br><span class="line"><span class="comment">Constructing...</span></span><br><span class="line"><span class="comment">WithDispose.dispose(): 1</span></span><br><span class="line"><span class="comment">Construction has failed: chapter12.exercise24.ConstructionException</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<h2 id="异常匹配"><a href="#异常匹配" class="headerlink" title="异常匹配"></a>异常匹配</h2><p>抛出异常的时候，异常处理系统会按照代码的书写顺序找出“最近”的处理程序，找到匹配的处理程序之后，它就认为异常将得到处理，然后就不再继续查找。</p>
<p>查找的时候并不要求抛出的异常同处理程序所声明的异常完全匹配。派生类的对象也可以匹配其基类的处理程序，就像这样：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">Level1Exception</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;&#125;</span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Level2Exception</span> <span class="keyword">extends</span> <span class="title class_">Level1Exception</span> &#123;&#125;</span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Level3Exception</span> <span class="keyword">extends</span> <span class="title class_">Level2Exception</span> &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">A</span> &#123;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> <span class="keyword">throws</span> Level1Exception &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Level1Exception</span>();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">B</span> <span class="keyword">extends</span> <span class="title class_">A</span> &#123;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> <span class="keyword">throws</span> Level2Exception &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Level2Exception</span>();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">class</span> <span class="title class_">C</span> <span class="keyword">extends</span> <span class="title class_">B</span> &#123;</span><br><span class="line">    <span class="keyword">void</span> <span class="title function_">f</span><span class="params">()</span> <span class="keyword">throws</span> Level3Exception &#123;</span><br><span class="line">        <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Level3Exception</span>();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">E25_ThreeLevelExceptions</span> &#123;</span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="type">A</span> <span class="variable">c</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">C</span>();</span><br><span class="line">            c.f();</span><br><span class="line">        <span class="comment">// catch按照异常的层级进行排列，基类异常在最后</span></span><br><span class="line">        &#125; <span class="keyword">catch</span> (Level3Exception l3) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Catch &quot;</span> + l3);</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Level2Exception l2) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Catch &quot;</span> + l2);</span><br><span class="line">        &#125; <span class="keyword">catch</span> (Level1Exception l1) &#123;</span><br><span class="line">            System.out.println(<span class="string">&quot;Catch &quot;</span> + l1);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;<span class="comment">/*Output:</span></span><br><span class="line"><span class="comment">// 找到最近的匹配处理程序之后，它就认为异常将得到处理，然后就不再继续查找。</span></span><br><span class="line"><span class="comment">Catch chapter12.exercise25.Level3Exception</span></span><br><span class="line"><span class="comment">*/</span></span><br></pre></td></tr></table></figure>

<h2 id="异常处理的思想及其他思考"><a href="#异常处理的思想及其他思考" class="headerlink" title="异常处理的思想及其他思考"></a>异常处理的思想及其他思考</h2><p>异常处理的一个重要原则是“只有在你知道如何处理的情况下才捕获异常”。实际上，异常处理的一个重要目标就是把错误处理的代码同错误发生的地点相分离。这使你能在一段代码中专注与要完成的事情，至于如何处理错误，则放在另一段代码中完成。这样一来，主干代码就不会与错误代码逻辑混在一起，也更容易理解和维护。通过允许一个处理程序去处理多个出错点，异常处理还使得错误处理代码的数量趋向于减少。</p>
<p>受查异常（checked exception）使得这个问题变得复杂，因为它们强制你在可能还没准备好处理错误时被迫加上catch子句，这就导致<strong>吞食则有害（harmfal if swallowed）</strong>的问题：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    <span class="comment">//...to do something useful</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(ObligatoryException e) &#123;&#125;  <span class="comment">// Gulp! Do nothing after catch it</span></span><br></pre></td></tr></table></figure>

<p>程序员们只做最简单的事情，常常是无意中“吞食”了异常，然而一且这么做，虽然能通过编译，但除非你记得复查并改正代码，否则异常将会丢失。异常确实发生了，但“吞食”后它却完全消失了。因为编译器强迫你立刻写代码来处理异常，所以这种看起来最简单的方法，却可能是最糟糕的做法。接下来将讨论一下<strong>受查异常及其并发症</strong>，以及采用什么方法解决这些问题。</p>
<p>仅从示意性的例子和小程序来看，“被检查的异常”的好处很明显。但是当程序开始变大的时候，就会带来一些微妙的问题。当然，程序不是一下就变大的，这有个过程。如果把不适用于大项目的语言用于小项目，当这些项目不断膨胀时，突然有一天你会发现，原来可以管理的东西，现在已经变得无法管理了。这就是我所说的过多的类型检查，特别是“被检查的异常”所造成的问题。</p>
<p>看来程序的规模是个重要因素。由于很多讨论都用小程序来做演示，因此这并不足以说明问题。一名C#的设计人员发现：“仅从小程序来看，会认为异常说明能增加开发人员的效率，并提高代码的质量；但考察大项目的时候，结论就不同了——开发效率下降了，而代码质量只有微不足道的提高，甚至毫无提高”。</p>
<p>然而，如果发现有些受查异常挡住了路，尤其是发现你不得不去对付那些不知道该如何处理的异常，还是有些办法的。</p>
<ol>
<li>把异常传递给控制台</li>
</ol>
<p>对于简单的程序，比如本书中的许多例子，最简单而又不用写多少代码就能保护异常信息的方法，就是把它们从main()传递到控制台。例如，为了读取信息而打开一个文件，必须对FileInputStream进行打开和关闭操作，这就可能会产生异常。对于简单的程序，可以像这样做：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">MainException</span> &#123;</span><br><span class="line">    <span class="comment">// Pass all exceptions to the console:</span></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> <span class="keyword">throws</span> Exception &#123;</span><br><span class="line">        <span class="comment">// Open the file:</span></span><br><span class="line">        <span class="type">FileInputStream</span> <span class="variable">file</span> <span class="operator">=</span> <span class="keyword">new</span> <span class="title class_">FileInputStream</span>(<span class="string">&quot;MainException, java&quot;</span>);</span><br><span class="line">        <span class="comment">// Use the file...</span></span><br><span class="line">        <span class="comment">// Close the file:</span></span><br><span class="line">         file.close();</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>注意， main()作为一个方法也可以有异常说明，这里异常的类型是 Exception,它也是所有受查异常的基类。通过把它传递到控制台，就不必在main()里写 try-catch子句了。</p>
<ol>
<li>把“受查异常”转换为“不受查异常”</li>
</ol>
<p>在编写你自己使用的简单程序时，从 main()中抛出异常是很方便的，但这不是通用的方法。异常链提供了一种新的思路来解决这个问题。可以直接把“受查异常”包装进 RuntimeException里面，就像这样：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    <span class="comment">//...to do something useful</span></span><br><span class="line">&#125; <span class="keyword">catch</span>(IDontKnowWhatToDoWithThisCheckedException e) &#123;</span><br><span class="line">    <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">RuntimeException</span>(e);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>如果想把“受查异常”这种功能“屏蔽”掉的话，这看上去像是一个好办法。不用“吞下”异常，也不必把它放到方法的异常说明里面，而异常链还能保证你不会丢失任何原始异常的信息。</p>
<p>这种技巧给了你一种选择，可以不写try-catch子句或异常说明，直接忽略异常，让他自己沿着调用栈往上“冒泡”，同时，还可以用getCause()捕获并处理特定的异常，就像这样：</p>
<figure class="highlight java"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">class</span> <span class="title class_">Annoyance</span> <span class="keyword">extends</span> <span class="title class_">Exception</span> &#123;&#125;</span><br><span class="line"><span class="keyword">class</span> <span class="title class_">Sneeze</span> <span class="keyword">extends</span> <span class="title class_">Annoyance</span> &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="keyword">public</span> <span class="keyword">class</span> <span class="title class_">E30_Human</span> &#123;</span><br><span class="line">    <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">throwRuntimeException</span><span class="params">(<span class="type">int</span> type)</span> &#123;</span><br><span class="line">        <span class="keyword">try</span> &#123;</span><br><span class="line">            <span class="keyword">switch</span>(type) &#123;</span><br><span class="line">                <span class="keyword">case</span> <span class="number">0</span>: <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Annoyance</span>();</span><br><span class="line">                <span class="keyword">case</span> <span class="number">1</span>: <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">Sneeze</span>();</span><br><span class="line">                <span class="keyword">default</span>: <span class="keyword">return</span>;</span><br><span class="line">            &#125;</span><br><span class="line">        &#125; <span class="keyword">catch</span>(Exception e) &#123;   <span class="comment">// 将受查异常转换成不受查异常:</span></span><br><span class="line">             <span class="keyword">throw</span> <span class="keyword">new</span> <span class="title class_">RuntimeException</span>(e);</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">public</span> <span class="keyword">static</span> <span class="keyword">void</span> <span class="title function_">main</span><span class="params">(String[] args)</span> &#123;</span><br><span class="line">         <span class="keyword">for</span>(<span class="type">int</span> <span class="variable">i</span> <span class="operator">=</span> <span class="number">0</span>; i &lt; <span class="number">2</span>; i++)</span><br><span class="line">             <span class="keyword">try</span> &#123;</span><br><span class="line">                 throwRuntimeException(i);  <span class="comment">// 抛出异常链</span></span><br><span class="line">             &#125; <span class="keyword">catch</span>(RuntimeException re) &#123;</span><br><span class="line">                 <span class="keyword">try</span> &#123;</span><br><span class="line">                     <span class="keyword">throw</span> re.getCause();   <span class="comment">// 获取原始异常</span></span><br><span class="line">                 &#125; <span class="keyword">catch</span>(Sneeze e) &#123;</span><br><span class="line">                     System.out.println(<span class="string">&quot;Caught Sneeze&quot;</span>);</span><br><span class="line">                 &#125; <span class="keyword">catch</span>(Annoyance e) &#123;</span><br><span class="line">                     System.out.println(<span class="string">&quot;Caught Annoyance&quot;</span>);</span><br><span class="line">                 &#125; <span class="keyword">catch</span>(Throwable t) &#123;</span><br><span class="line">                     System.out.println(t);</span><br><span class="line">                 &#125;</span><br><span class="line">             &#125;</span><br><span class="line">    &#125;</span><br><span class="line">&#125; <span class="comment">/* Output:</span></span><br><span class="line"><span class="comment">Caught Annoyance</span></span><br><span class="line"><span class="comment">Caught Sneeze</span></span><br><span class="line"><span class="comment">*/</span><span class="comment">//:~</span></span><br></pre></td></tr></table></figure>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="http://example.com">WangHao</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="http://example.com/2021/04/29/%E5%BC%82%E5%B8%B8/">http://example.com/2021/04/29/%E5%BC%82%E5%B8%B8/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="http://example.com" target="_blank">WangHao&博客</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E5%BC%82%E5%B8%B8/">异常</a></div><div class="post_share"><div class="social-share" data-image="/img/40.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2022/05/12/JVM%E5%86%85%E5%AD%98%E5%8C%BA%E5%9F%9F/"><img class="prev-cover" src="/img/60.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">JVM内存区域</div></div></a></div><div class="next-post pull-right"><a href="/2021/04/25/%E5%AE%B9%E5%99%A8/"><img class="next-cover" src="/img/35.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">容器</div></div></a></div></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/img/head.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">WangHao</div><div class="author-info__description">Stay hungry,stay foolish</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">101</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">120</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">13</div></a></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E6%A6%82%E5%BF%B5"><span class="toc-number">1.</span> <span class="toc-text">异常概念</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E5%88%86%E7%B1%BB%EF%BC%9A%E4%B8%A4%E7%A7%8D%E5%88%86%E7%B1%BB%E6%96%B9%E6%B3%95"><span class="toc-number">2.</span> <span class="toc-text">异常分类：两种分类方法</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#Error-%E5%92%8C-Exception"><span class="toc-number">2.1.</span> <span class="toc-text">Error 和 Exception</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%97%E6%9F%A5%E5%BC%82%E5%B8%B8%E5%92%8C%E4%B8%8D%E5%8F%97%E6%9F%A5%E5%BC%82%E5%B8%B8"><span class="toc-number">2.2.</span> <span class="toc-text">受查异常和不受查异常</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E6%9C%BA%E5%88%B6"><span class="toc-number">3.</span> <span class="toc-text">异常处理机制</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8D%95%E8%8E%B7%E4%B8%8E%E5%A4%84%E7%90%86"><span class="toc-number">3.1.</span> <span class="toc-text">捕获与处理</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#try-catch-finally%E7%9A%84%E5%A4%84%E7%90%86%E6%9C%BA%E5%88%B6"><span class="toc-number">3.1.1.</span> <span class="toc-text">try-catch-finally的处理机制</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%87%AA%E5%8A%A8%E5%85%B3%E9%97%AD%E7%9A%84try-catch"><span class="toc-number">3.1.2.</span> <span class="toc-text">自动关闭的try-catch</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#try-catch-finally%E7%9A%84%E6%89%A7%E8%A1%8C%E6%9C%BA%E5%88%B6"><span class="toc-number">3.1.3.</span> <span class="toc-text">try-catch-finally的执行机制</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E4%B8%A2%E5%A4%B1"><span class="toc-number">3.1.4.</span> <span class="toc-text">异常丢失</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%8D%95%E8%8E%B7%E6%89%80%E6%9C%89%E5%BC%82%E5%B8%B8"><span class="toc-number">3.2.</span> <span class="toc-text">捕获所有异常</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E4%BF%A1%E6%81%AF%EF%BC%9Ae-toString-%E5%92%8Ce-getMessage"><span class="toc-number">3.2.1.</span> <span class="toc-text">异常的信息：e.toString()和e.getMessage()</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A0%88%E8%BD%A8%E8%BF%B9%E5%92%8C%E9%87%8D%E6%8A%9B%E5%BC%82%E5%B8%B8"><span class="toc-number">3.2.2.</span> <span class="toc-text">栈轨迹和重抛异常</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#e-printStackTrace"><span class="toc-number">3.2.2.1.</span> <span class="toc-text">e.printStackTrace()</span></a></li><li class="toc-item toc-level-4"><a class="toc-link" href="#%E9%87%8D%E6%96%B0%E6%8A%9B%E5%87%BA%E5%BC%82%E5%B8%B8"><span class="toc-number">3.2.2.2.</span> <span class="toc-text">重新抛出异常</span></a></li></ol></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E9%93%BE"><span class="toc-number">3.2.3.</span> <span class="toc-text">异常链</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8"><span class="toc-number">4.</span> <span class="toc-text">自定义异常</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%9B%E5%BB%BA%E8%87%AA%E5%AE%9A%E4%B9%89%E5%BC%82%E5%B8%B8"><span class="toc-number">4.1.</span> <span class="toc-text">创建自定义异常</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E8%AF%B4%E6%98%8E"><span class="toc-number">4.2.</span> <span class="toc-text">异常说明</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E9%99%90%E5%88%B6"><span class="toc-number">4.3.</span> <span class="toc-text">异常的限制</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E4%B8%8E%E8%AE%B0%E5%BD%95%E6%97%A5%E5%BF%97"><span class="toc-number">4.4.</span> <span class="toc-text">异常与记录日志</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#System-err%E5%92%8CSystem-out%E7%9A%84%E5%8C%BA%E5%88%AB%EF%BC%88%E8%A1%A5%E5%85%85%EF%BC%89"><span class="toc-number">4.4.1.</span> <span class="toc-text">System.err和System.out的区别（补充）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Java-Logger%E4%BD%BF%E7%94%A8"><span class="toc-number">4.4.2.</span> <span class="toc-text">Java Logger使用</span></a></li></ol></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E6%B7%B1%E5%85%A5"><span class="toc-number">5.</span> <span class="toc-text">异常深入</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%9E%84%E9%80%A0%E5%99%A8"><span class="toc-number">5.1.</span> <span class="toc-text">构造器</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E5%8C%B9%E9%85%8D"><span class="toc-number">5.2.</span> <span class="toc-text">异常匹配</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E7%9A%84%E6%80%9D%E6%83%B3%E5%8F%8A%E5%85%B6%E4%BB%96%E6%80%9D%E8%80%83"><span class="toc-number">5.3.</span> <span class="toc-text">异常处理的思想及其他思考</span></a></li></ol></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>最新文章</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2023/01/06/RocketMQ%E6%B6%88%E6%81%AF%E5%8F%AF%E9%9D%A0%E6%80%A7%E4%BF%9D%E8%AF%81/" title="RocketMQ消息可靠性保证"><img src="/img/100.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="RocketMQ消息可靠性保证"/></a><div class="content"><a class="title" href="/2023/01/06/RocketMQ%E6%B6%88%E6%81%AF%E5%8F%AF%E9%9D%A0%E6%80%A7%E4%BF%9D%E8%AF%81/" title="RocketMQ消息可靠性保证">RocketMQ消息可靠性保证</a><time datetime="2023-01-06T14:57:15.000Z" title="发表于 2023-01-06 22:57:15">2023-01-06</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/01/04/RocketMQ%E8%AF%A6%E8%A7%A3%E4%B8%8E%E9%A1%B9%E7%9B%AE%E4%BD%BF%E7%94%A8/" title="RocketMQ详解与项目使用"><img src="/img/default.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="RocketMQ详解与项目使用"/></a><div class="content"><a class="title" href="/2023/01/04/RocketMQ%E8%AF%A6%E8%A7%A3%E4%B8%8E%E9%A1%B9%E7%9B%AE%E4%BD%BF%E7%94%A8/" title="RocketMQ详解与项目使用">RocketMQ详解与项目使用</a><time datetime="2023-01-04T08:24:40.000Z" title="发表于 2023-01-04 16:24:40">2023-01-04</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/01/02/Spring%E4%BA%8B%E5%8A%A1/" title="Spring事务"><img src="/img/99.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Spring事务"/></a><div class="content"><a class="title" href="/2023/01/02/Spring%E4%BA%8B%E5%8A%A1/" title="Spring事务">Spring事务</a><time datetime="2023-01-02T15:46:49.000Z" title="发表于 2023-01-02 23:46:49">2023-01-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/01/02/SpringBoot%E8%87%AA%E5%8A%A8%E8%A3%85%E9%85%8D%E5%8E%9F%E7%90%86/" title="SpringBoot自动装配原理"><img src="/img/98.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="SpringBoot自动装配原理"/></a><div class="content"><a class="title" href="/2023/01/02/SpringBoot%E8%87%AA%E5%8A%A8%E8%A3%85%E9%85%8D%E5%8E%9F%E7%90%86/" title="SpringBoot自动装配原理">SpringBoot自动装配原理</a><time datetime="2023-01-02T14:57:28.000Z" title="发表于 2023-01-02 22:57:28">2023-01-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2023/01/02/SpringBoot%E7%9A%84%E5%90%AF%E5%8A%A8%E6%B5%81%E7%A8%8B/" title="SpringBoot的启动流程"><img src="/img/95.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="SpringBoot的启动流程"/></a><div class="content"><a class="title" href="/2023/01/02/SpringBoot%E7%9A%84%E5%90%AF%E5%8A%A8%E6%B5%81%E7%A8%8B/" title="SpringBoot的启动流程">SpringBoot的启动流程</a><time datetime="2023-01-02T12:40:49.000Z" title="发表于 2023-01-02 20:40:49">2023-01-02</time></div></div></div></div></div></div></main><footer id="footer" style="background-image: url('/img/40.jpg')"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2023 By WangHao</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><div class="js-pjax"></div><canvas class="fireworks" mobile="false"></canvas><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/fireworks.min.js"></script><script id="canvas_nest" defer="defer" color="0,0,255" opacity="0.7" zIndex="-1" count="99" mobile="false" src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/canvas-nest.min.js"></script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div></body></html>